hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java

```diff
@@ -211,11 +212,12 @@ long uploadJobResources(JobConfig jobConfig) {
                         }
                         break;
                     case FILE:
-                        try (
-                                InputStream in = rc.getUrl().openStream();
-                                IMapOutputStream os = new IMapOutputStream(jobFileStorage.get(), fileKeyName(rc.getId()))
+                        Path fnamePath = Paths.get(Objects.requireNonNull(rc.getUrl().getPath())).getFileName();
+                        assert fnamePath != null; // needed to silence SpotBugs
+                        try (InputStream in = rc.getUrl().openStream();
+                             IMapOutputStream os = new IMapOutputStream(jobFileStorage.get(), fileKeyName(rc.getId()))
                         ) {
-                            packStreamIntoZip(in, os, rc.getId());
+                            packStreamIntoZip(in, os, fnamePath.toString());
```

Comment:
This is rather confusing. If the user doesn't use the `attachFile` variant with `id`, then the file name and ID are always equal. If he does, then he does so for some reason. We should use the ID he asked for and not the one from the URL. The URL might even not contain a file name, but we require it here.

If we keep it like this, we should document at `attachFile(URL, id)` that we require the URL to contain a file name and that that file name will be the name of the file, the ID is just used to look up that file.

Reply:
ID must be unique, and if we use it as the filename then no two resources can have the same filename. On the other hand, a given filename may be required and there is no uniqueness requirement.

Reply:
Then we should document that the URL must contain a file name and that the file will actually have that name and ID is used to query that file. We can also mention that multiple files can have the same name and different ID, it will make that option obvious.

Reply:
The URL has in fact always had an even stronger requirement: to be resolvable into a local file. We should document that explicitly.

Reply:
Maybe in the past, now we only use `url.openStream()`, so it has to be any URL that can be read.

Reply:
Maybe in the past, but now we only use `url.openStream()`

Reply:
The complete story is in `JobConfig`, where we convert URLs to files. I'm adding more explicit checking there.

Reply:
We convert the URLs to file names, not to files. A http URL works just fine, if it contains a file name.

Nevertheless, I was thinking about it more. Maybe we can leave it as it is. It's very unlikely to encounter a URL like `http://hazelcast.com/` that has no file name. The user will receive a `NPE` or `AssertionError` in this case, which is very easy to fix by looking at the code. I'm marking this as resolved.

Reply:
In the meantime i cleaned up `JobConfig` to make the checks where needed. `attachFile/Directory` must get a `file` URL and checks for the existence of the file. The `add` variants can get an opaque URL that they don't check, but they still check it has at least one path segment to be used as the ID.